File: macos/Sources/ChoroReader/Search/DocumentSearch.swift
import Foundation
import PDFKit

struct SearchResult: Identifiable, Hashable {
    let id = UUID()
    var locator: Locator
    var chapterTitle: String
    var before: String
    var match: String
    var after: String
    var isCode: Bool
}

/// 章の HTML から本文とコードを取り出す。索引は作らず、要求されたときに走査する。
/// 全文インデックス（SQLite FTS）は第 2 段階。
enum HTMLText {
    struct Extracted {
        var text: String
        /// コードブロックが占める範囲。検索結果の種別表示に使う。
        var codeRanges: [Range<Int>]

        func isCode(at offset: Int) -> Bool {
            codeRanges.contains { $0.contains(offset) }
        }
    }

    private static let entities: [String: String] = [
        "&amp;": "&", "&lt;": "<", "&gt;": ">", "&quot;": "\"", "&apos;": "'",
        "&nbsp;": " ", "&mdash;": "—", "&ndash;": "–", "&hellip;": "…",
    ]

    static func extract(_ html: String) -> Extracted {
        var source = html
        source = remove(pattern: #"(?is)<script\b[^>]*>.*?</script>"#, from: source)
        source = remove(pattern: #"(?is)<style\b[^>]*>.*?</style>"#, from: source)
        source = remove(pattern: #"(?s)<!--.*?-->"#, from: source)

        var text = ""
        var codeRanges: [Range<Int>] = []

        guard let re = try? NSRegularExpression(pattern: #"(?is)<pre\b[^>]*>(.*?)</pre>"#) else {
            return Extracted(text: stripTags(source), codeRanges: [])
        }

        var cursor = source.startIndex
        let full = NSRange(source.startIndex..., in: source)
        for match in re.matches(in: source, range: full) {
            guard let whole = Range(match.range, in: source),
                  let inner = Range(match.range(at: 1), in: source) else { continue }
            text += stripTags(String(source[cursor ..< whole.lowerBound]))
            let start = text.count
            text += stripTags(String(source[inner]))
            codeRanges.append(start ..< text.count)
            text += "\n"
            cursor = whole.upperBound
        }
        text += stripTags(String(source[cursor...]))

        return Extracted(text: text, codeRanges: codeRanges)
    }

    private static func remove(pattern: String, from s: String) -> String {
        guard let re = try? NSRegularExpression(pattern: pattern) else { return s }
        return re.stringByReplacingMatches(in: s, range: NSRange(s.startIndex..., in: s), withTemplate: "")
    }

    static func stripTags(_ s: String) -> String {
        var out = ""
        out.reserveCapacity(s.count)
        var insideTag = false
        var lastWasSpace = false
        for ch in s {
            if ch == "<" { insideTag = true; continue }
            if ch == ">" {
                insideTag = false
                if !lastWasSpace { out.append(" "); lastWasSpace = true }
                continue
            }
            if insideTag { continue }
            if ch == "\n" || ch == "\r" || ch == "\t" {
                if !lastWasSpace { out.append(" "); lastWasSpace = true }
                continue
            }
            out.append(ch)
            lastWasSpace = (ch == " ")
        }
        return decodeEntities(out)
    }

    private static func decodeEntities(_ s: String) -> String {
        guard s.contains("&") else { return s }
        var out = s
        for (k, v) in entities { out = out.replacingOccurrences(of: k, with: v) }
        guard out.contains("&#"), let re = try? NSRegularExpression(pattern: #"&#(x?)([0-9A-Fa-f]+);"#) else {
            return out
        }
        var result = out
        for m in re.matches(in: out, range: NSRange(out.startIndex..., in: out)).reversed() {
            guard let whole = Range(m.range, in: out),
                  let flagRange = Range(m.range(at: 1), in: out),
                  let digitsRange = Range(m.range(at: 2), in: out) else { continue }
            let radix = out[flagRange].isEmpty ? 10 : 16
            guard let code = UInt32(out[digitsRange], radix: radix), let scalar = Unicode.Scalar(code) else { continue }
            result.replaceSubrange(whole, with: String(Character(scalar)))
        }
        return result
    }
}

enum DocumentSearch {
    static let resultLimit = 400
    /// 日本語では単語境界が定まらないため、標準は部分一致とする。全角と半角も区別しない。
    static let options: String.CompareOptions = [.caseInsensitive, .widthInsensitive, .diacriticInsensitive]

    struct Outcome {
        var results: [SearchResult]
        var truncated: Bool
    }

    static func searchEPUB(resources: ResourceProvider, publication: EPUBPublication, query: String,
                           completion: @escaping (Outcome) -> Void) {
        let order = publication.readingOrder
        let titles = order.map { publication.title(forHref: $0.href) ?? ($0.href as NSString).lastPathComponent }

        DispatchQueue.global(qos: .userInitiated).async {
            var results: [SearchResult] = []
            var truncated = false

            outer: for (i, link) in order.enumerated() {
                guard let data = try? resources.read(link.href) else { continue }
                let extracted = HTMLText.extract(CSSCompat.decodeText(data))
                let text = extracted.text
                guard !text.isEmpty else { continue }
                let chars = Array(text)

                var searchStart = text.startIndex
                while let range = text.range(of: query, options: options, range: searchStart ..< text.endIndex) {
                    let offset = text.distance(from: text.startIndex, to: range.lowerBound)
                    let matchLength = text.distance(from: range.lowerBound, to: range.upperBound)
                    let before = String(chars[max(0, offset - 30) ..< offset])
                    let after = String(chars[min(chars.count, offset + matchLength) ..< min(chars.count, offset + matchLength + 40)])

                    results.append(SearchResult(
                        locator: Locator(href: link.href,
                                         progression: chars.isEmpty ? 0 : Double(offset) / Double(chars.count),
                                         title: titles[i],
                                         text: String(chars[offset ..< min(chars.count, offset + max(matchLength, 12))])),
                        chapterTitle: titles[i],
                        before: before.trimmingCharacters(in: .whitespaces),
                        match: String(text[range]),
                        after: after.trimmingCharacters(in: .whitespaces),
                        isCode: extracted.isCode(at: offset)
                    ))

                    if results.count >= resultLimit { truncated = true; break outer }
                    searchStart = range.upperBound
                }
            }

            let outcome = Outcome(results: results, truncated: truncated)
            DispatchQueue.main.async { completion(outcome) }
        }
    }
}

/// PDF は PDFKit の非同期検索を使い、UI を止めない。
@MainActor
final class PDFSearcher {
    private var observers: [NSObjectProtocol] = []
    private var collected: [SearchResult] = []
    private var onUpdate: (([SearchResult], Bool) -> Void)?

    func cancel() {
        observers.forEach { NotificationCenter.default.removeObserver($0) }
        observers = []
    }

    func search(document: PDFKit.PDFDocument, query: String,
                onUpdate: @escaping ([SearchResult], Bool) -> Void) {
        cancel()
        collected = []
        self.onUpdate = onUpdate
        if document.isFinding { document.cancelFindString() }

        let center = NotificationCenter.default
        observers.append(center.addObserver(forName: .PDFDocumentDidFindMatch, object: document,
                                            queue: .main) { [weak self] note in
            MainActor.assumeIsolated {
                guard let self, let selection = note.userInfo?["PDFDocumentFoundSelection"] as? PDFSelection else { return }
                self.append(selection, in: document)
            }
        })
        observers.append(center.addObserver(forName: .PDFDocumentDidEndFind, object: document,
                                            queue: .main) { [weak self] _ in
            MainActor.assumeIsolated {
                guard let self else { return }
                self.onUpdate?(self.collected, self.collected.count >= DocumentSearch.resultLimit)
                self.cancel()
            }
        })

        document.beginFindString(query, withOptions: DocumentSearch.options)
    }

    private func append(_ selection: PDFSelection, in document: PDFKit.PDFDocument) {
        guard collected.count < DocumentSearch.resultLimit,
              let page = selection.pages.first else { return }
        let index = document.index(for: page)
        let context = contextText(for: selection, on: page)
        collected.append(SearchResult(
            locator: Locator(page: index,
                             progression: document.pageCount > 1 ? Double(index) / Double(document.pageCount - 1) : 0,
                             title: page.label.map { "p.\($0)" },
                             text: selection.string),
            chapterTitle: "ページ \(page.label ?? String(index + 1))",
            before: context.before,
            match: selection.string ?? "",
            after: context.after,
            isCode: false
        ))
        if collected.count % 20 == 0 { onUpdate?(collected, false) }
    }

    private func contextText(for selection: PDFSelection, on page: PDFPage) -> (before: String, after: String) {
        guard let pageText = page.string, let match = selection.string,
              let range = pageText.range(of: match) else { return ("", "") }
        let chars = Array(pageText)
        let offset = pageText.distance(from: pageText.startIndex, to: range.lowerBound)
        let end = offset + match.count
        let before = String(chars[max(0, offset - 30) ..< offset])
        let after = String(chars[min(chars.count, end) ..< min(chars.count, end + 40)])
        return (before.trimmingCharacters(in: .whitespacesAndNewlines),
                after.trimmingCharacters(in: .whitespacesAndNewlines))
    }
}

File: windows/ChoroReader.Core/DocumentSearch.cs
using System.Text;

namespace ChoroReader.Core;

/// <param name="Nth">
/// 読み順の 1 項目の中で何番目の当たりか。
/// 飛んだ先で「押したのはこの語のどれか」を選び直すために使う。
/// </param>
public sealed record SearchResult(
    Locator Locator,
    string ChapterTitle,
    string Before,
    string Match,
    string After,
    bool IsCode,
    int Nth);

public sealed record SearchOutcome(IReadOnlyList<SearchResult> Results, bool Truncated);

/// <summary>章を順に走査して照合位置を返す。索引は持たない。</summary>
public static class DocumentSearch
{
    public const int ResultLimit = 400;

    public static SearchOutcome SearchEpub(
        IResourceProvider resources, EpubPublication publication, string query) =>
        SearchEpubWithin(resources, publication, query, null, ResultLimit);

    /// <summary>
    /// 読み順のうち <paramref name="only"/> に挙がった位置だけを走査する。null なら全部。
    /// </summary>
    public static SearchOutcome SearchEpubWithin(
        IResourceProvider resources,
        EpubPublication publication,
        string query,
        IReadOnlyCollection<int>? only,
        int limit)
    {
        var results = new List<SearchResult>();
        var truncated = false;

        var needle = Fold.Text(query);
        if (needle.Count == 0)
        {
            return new SearchOutcome(results, truncated);
        }

        for (var position = 0; position < publication.ReadingOrder.Count; position++)
        {
            if (truncated)
            {
                break;
            }
            if (only is not null && !only.Contains(position))
            {
                continue;
            }
            var link = publication.ReadingOrder[position];
            var source = resources.ReadText(link.Href);
            if (source is null)
            {
                continue;
            }

            var title = publication.TitleForHref(link.Href) ?? EpubParser.LastComponent(link.Href);
            var extracted = HtmlText.Extract(source);
            var runes = Runes(extracted.Text);
            if (runes.Count == 0)
            {
                continue;
            }

            // 通し番号は読み順の項目ごとに数え直す。同じ経路が読み順に 2 度出ることがあり、
            // そのときは同じ文書を 2 度開くので、番号も 0 から振り直さないと選び直せない。
            var nth = 0;
            foreach (var (found, matchEnd) in Matches(runes, needle))
            {
                var snippetEnd = Math.Min(runes.Count, found + Math.Max(matchEnd - found, 12));

                results.Add(new SearchResult(
                    new Locator
                    {
                        Href = link.Href,
                        Progression = (double)found / runes.Count,
                        Title = title,
                        Text = Slice(runes, found, snippetEnd),
                    },
                    title,
                    Slice(runes, Math.Max(0, found - 30), found).Trim(),
                    Slice(runes, found, matchEnd),
                    Slice(runes, matchEnd, Math.Min(runes.Count, matchEnd + 40)).Trim(),
                    extracted.IsCode(found),
                    nth));
                nth++;

                if (results.Count >= limit)
                {
                    truncated = true;
                    break;
                }
            }
        }

        return new SearchOutcome(results, truncated);
    }

    /// <summary>
    /// 本文の中で nth 番目の当たりが占める範囲（文字単位、終わりは含まない）。
    ///
    /// 走査と同じ畳み方で数えるので、検索が返した通し番号と、ここで選ぶ当たりは必ず一致する。
    /// 当たりを強調するとき、どの語を囲むかをこれで決める。
    /// </summary>
    public static (int From, int To)? NthMatch(string text, string query, int nth)
    {
        foreach (var found in Matches(Runes(text), Fold.Text(query)))
        {
            if (nth-- == 0)
            {
                return found;
            }
        }
        return null;
    }

    /// <summary>
    /// 本文に当たる語が現れる範囲を、元の文字位置で順に返す。
    ///
    /// 走査も強調も、当たりを数えるのはここ 1 か所にする。
    /// 別々に数えると、畳み方を変えたときに片方だけ直り、
    /// 検索が言う「何番目」と、強調が囲む語が食い違う。
    /// </summary>
    private static IEnumerable<(int From, int To)> Matches(
        IReadOnlyList<Rune> runes, IReadOnlyList<Rune> needle)
    {
        if (needle.Count == 0)
        {
            yield break;
        }
        var haystack = Fold.All(runes);
        var total = runes.Count;
        var from = 0;

        while (from + needle.Count <= haystack.Folded.Count)
        {
            var hit = Find(haystack.Folded, from, needle);
            if (hit < 0)
            {
                yield break;
            }
            var found = haystack.Origin[hit];
            // 照合の長さは元の文字列側で数える。畳んだ結果と長さが変わりうるため。
            var length = OriginalLength(haystack, hit, needle.Count, total);
            from = hit + 1;
            yield return (found, Math.Min(found + length, total));
        }
    }

    /// <summary>畳んだ列の何文字ぶんが、元の何文字に当たるかを求める。</summary>
    private static int OriginalLength(Fold.Folding haystack, int foldedAt, int needleLength, int total)
    {
        var start = haystack.Origin[foldedAt];
        var end = foldedAt + needleLength < haystack.Origin.Count
            ? haystack.Origin[foldedAt + needleLength]
            : total;
        return Math.Max(end - start, 1);
    }

    /// <summary><paramref name="from"/> 以降で最初に一致する位置。見つからなければ -1。</summary>
    private static int Find(IReadOnlyList<Rune> haystack, int from, IReadOnlyList<Rune> needle)
    {
        for (var start = from; start + needle.Count <= haystack.Count; start++)
        {
            var same = true;
            for (var offset = 0; offset < needle.Count; offset++)
            {
                if (haystack[start + offset] != needle[offset])
                {
                    same = false;
                    break;
                }
            }
            if (same)
            {
                return start;
            }
        }
        return -1;
    }

    public static List<Rune> Runes(string text)
    {
        var runes = new List<Rune>(text.Length);
        foreach (var rune in text.EnumerateRunes())
        {
            runes.Add(rune);
        }
        return runes;
    }

    private static string Slice(IReadOnlyList<Rune> runes, int start, int end)
    {
        start = Math.Min(start, runes.Count);
        end = Math.Clamp(end, start, runes.Count);
        var text = new StringBuilder(end - start);
        for (var index = start; index < end; index++)
        {
            text.Append(runes[index].ToString());
        }
        return text.ToString();
    }
}
